cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsTableMigration.java

```diff
@@ -0,0 +1,206 @@
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsTableMigration.class);
+
+  private final MetricsTable v2MetricsTable;
+  private final MetricsTable v3MetricsTable;
+  private final Configuration hConf;
+  private final CConfiguration cConf;
+
+  public MetricsTableMigration(MetricDatasetFactory metricDatasetFactory, int resolution,
+                               CConfiguration cConf, Configuration hConf) {
+    this(metricDatasetFactory.getV2MetricsTable(resolution),
+         metricDatasetFactory.getV3MetricsTable(resolution), cConf, hConf);
+  }
+
+  public MetricsTableMigration(MetricsTable v2MetricsTable, MetricsTable v3MetricsTable,
+                               CConfiguration cConf, Configuration hConf) {
+    this.v2MetricsTable = v2MetricsTable;
+    this.v3MetricsTable = v3MetricsTable;
+    this.cConf = cConf;
+    this.hConf = hConf;
+  }
+
+  /**
+   * returns true if old metrics data is still available for transfer
+   * @return
+   */
+  public boolean isOldMetricsDataAvailable() {
+    Scanner scanner = v2MetricsTable.scan(null, null, null);
+    Row row = scanner.next();
+    scanner.close();
+    return (row != null);
+  }
+
+  /**
+   * transfers data from v2 metrics table to v3 metrics table, limit on number of records is specified by
+   * maxRecordsToScan
+   * @param maxRecordsToScan - limit on number of records scanned
+   */
+  public void transferData(int maxRecordsToScan) {
+    LOG.trace("Running Metrics Migrator Thread");
+    Scanner scanner = v2MetricsTable.scan(null, null, null);
+    Row row;
+    int recordsScanned = 0;
+
+    while (((row = scanner.next()) != null) && (recordsScanned < maxRecordsToScan)) {
+      if (recordsScanned % 10 == 0) {
+        LOG.trace("Scanned {} records in Metrics Data Migration", recordsScanned);
+      }
+
+      byte[] rowKey = row.getRow();
+      Map<byte[], byte[]> columns = row.getColumns();
+      //row-map
+      SortedMap<byte[], SortedMap<byte[], Long>> rowGauges = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+      NavigableMap<byte[], NavigableMap<byte[], Long>> rowIncrements = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+
+      // column-map gauges
+      SortedMap<byte[], Long> gauges = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+      // column-map increments
+      NavigableMap<byte[], Long> increments = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+
+      for (Map.Entry<byte[], byte[]> entry : columns.entrySet()) {
+        // column is timestamp, do a get on the new table
+        byte[] value = v3MetricsTable.get(rowKey, entry.getKey());
+        if (value == null) {
+          gauges.put(entry.getKey(), Bytes.toLong(entry.getValue()));
+        } else {
+          increments.put(entry.getKey(), Bytes.toLong(entry.getValue()));
+        }
+      }
+      byte[][] deletes = getByteArrayFromSets(increments.keySet(), gauges.keySet());
+
+      if (deletes.length > 0) {
+        v2MetricsTable.delete(rowKey, deletes);
+      }
+
+      if (!gauges.isEmpty()) {
```

Comment:
for the gauges, you need to prevent a conflict with the metrics processor. It might write a value to the v3 table (and delete the one in v2, or keep in in v2 if it is actually an increment) between the time you read the row and and the time you write the gauges to v3. Therefore you need to use checkAndPut, and retry if the v3 value has changed.